Prompt cancellation

Aidan63 opened the pull request:
Closes #105.

There are several parts to this, first the schedulers `schedule` function returns a handle which allows you to stop that function from being executed in the future. If it's currently being executed or has already been it does nothing. This is completely independent and unrelated to the cancellation propagation of structured concurrency.

The second is `ICancellationToken` interface which `AbstractTask` now implements. This exposes a read only property for if cancellation has been requested and the ability to register a callback which will be registered if / when the task is cancelled. These two things allow delay to now exit immediately when cancellation occurs.

I ultimately went with a cancellation token approach here as now that we're thinking of splitting hxcoro into a haxelib I think there is good value in a basic "universal" cancellation mechanism in the coro std lib. Thinking of something like asys which would assumably be in the std lib, if it were to support cancellation then it could use the cancellation token interfaces to do so. hxcoro would then work with it and if someone else makes a coroutine framework all they need to do is also support that basic interface. Hopefully it's slim enough to be unobtrusive to implement but also actually useful (polling and callback seems like it will cover all bases).

There are still a few things I want to do / decide on.

- It would be nice if there was a cancellation token key for the context, even if it ultimately returns the same task object. But I don't think that possible with our current approach.
- I wrote some (hopefully useful) comments on some of the interfaces, I want to do the same for cancellation token as well and will try and do similar going forward so save the pain of having to do it all at the end.
- I'm being quite eager in allocating the empty arrays for tasks which will hold the handles. This could probably be done lazilly to stop allocations if the api is not used.

Simn commented on std/hxcoro/AbstractTask.hx:
```diff
@@ -21,21 +62,36 @@ class TaskException extends Exception {}
+abstract class AbstractTask<T> implements ICancellationToken {
 	final children:Array<AbstractTask<Any>>;
+	final cancellationCallbacks:Array<CancellationHandle>;
+	final noOpCancellationHandle:NoOpCancellationHandle;
```
This looks like it should become a static.

Aidan63 replied:
I've moved it into a static.

Simn commented on std/hxcoro/AbstractTask.hx:
```diff
@@ -81,15 +142,18 @@ abstract class AbstractTask<T> {
-	**/
-	public function cancellationRequested() {
-		return switch (state) {
+	public function onCancellationRequested(f:()->Void):ICancellationHandle {
```
I've been making an effort to avoid storing closures in tasks because I would like to keep serialization a possibility. Plus, in my mind arbitrary callbacks like this diminish the whole "structured" aspect we're going for here.

It might be worthwhile to introduce some kind of `ICloseable` interface for this, though a the same time I'm worried about turning everything too Java.

Aidan63 replied:
I'm not sure there's a way around storing closures in this case, we want some way to be able to react immediately to cancellation. I'm hoping it doesn't break structured concurrency since it's part of the cancellation token interface and exposes none of the task internals.
I also thought about adding something like `IClosable` or `IHandle` since there are now a few interfaces which just have a `close` function.

Simn replied:
Wouldn't something like this work?

```haxe
class CloseAndCancel<T> implements ICloseable {
	public final handle:ICloseable;
	public final cont:IContinuation<T>;

	public function new(handle:ICloseable, cont:IContinuation<T>) {
		this.handle = handle;
		this.cont = cont;
	}

	public function close() {
		handle.close();
		cont.resume(null, new CancellationException());
	}
}
```

And then we register this as an `ICloseable` on the task, which can either call its `close()` immediately or store the object for later.

Aidan63 replied:
I think I'm getting myself in a muddle. In this case would `onCancellationRequested` accept a `IContinuation` instead of a function? In that sample is the `handle` field a handle from a scheduler?

Simn replied:
Sorry for throwing random pieces of code at you, I should probably just implement it properly instead and see if it works!

```haxe
cancellationHandle = ct.onCancellationRequested(new CloseAndCancel(scheduleHandle, cont))
```

And then:

```haxe
	public function onCancellationRequested(closeable:ICloseable):ICancellationHandle {
		return switch state {
			case Cancelling | Cancelled:
				closeable.close();

				return noOpCancellationHandle;
			case _:
				final container = cancellationCallbacks ??= [];
				final handle    = new CancellationHandle(closeable, container);

				container.push(handle);

				handle;
		}
	}
```

And also have `CancellationHandle` call `func.close()` instead.

Aidan63 replied:
I see now. This would mean that any sort of custom functionality you needed with prompt cancellation would require a custom `IClosable` handle. E.g. with the new channel there would need to be closables which removes suspending read and writes from the queue if one of them was cancelled. Hmm.

Simn replied:
We don't have to do this right now if it makes your life harder. I certainly like how the closure version looks because it's quite direct and readable, so I'm fine with sticking to it for now. If it gets in the way later, we can still change it.

Aidan63 replied:
Sounds good. We might get a better idea of how it should work once we start using it in non-trivial ways.

Simn commented:
> It would be nice if there was a cancellation token key for the context, even if it ultimately returns the same task object. But I don't think that possible with our current approach.

I came across this too, but it's a simple matter of an additional context function which I've added in https://github.com/Aidan63/haxe/commit/239e014e2ffce5efbc80e11c613b0cf22a31d592. This allows `context.clone().with(this).add(CancellationToken.key, this)`.

> I'm being quite eager in allocating the empty arrays for tasks which will hold the handles. This could probably be done lazilly to stop allocations if the api is not used.

That's something else I was also thinking about in the context of `AbstractTask.children`, which could be lazily initialized as well.

Aidan63 commented:
I've added a cancellation token key and updated delay and yield to use that, the children and callbacks arrays are also lazilly allocated in abstract task.

In the `addChild` function I tried to get a bit fancy and use those new null operators, but got a "Invalid assign" compiler error when trying to do this.

```haxe
final container = children ?? children = [];
final index     = container.push(child);

child.indexInParent = index - 1;
```

Is that not allowed? It looks like it should be (to me at least), in the end I went back to a bog standard null if check.

Simn commented:
That's probably parsed as `(children ?? children) = []` due to operator precedence.

RblSb commented:
This also can be written as `container = children ??= [];`

Aidan63 commented:
That makes sense, didn't know about that syntax, thanks!
